Frameworks/CoreGraphics/CGImage.mm

```diff
@@ -786,4 +788,83 @@ REFGUID _CGImageGetWICPixelFormatFromImageProperties(unsigned int bitsPerCompone
     return GUID_WICPixelFormat32bppRGBA;
 }
 
+// __CGImageMaskConvertToWICAlphaBitmap converts a 1, 2, 4, or 8-bpp grayscale "mask" into an alpha-only image for a D2D opacity brush.
+// On the reference platform, mask images are grayscale images whose pixel values signify the pixel's alpha transmissivity.
+// A fully black region (S = 0.0) is translated to a fully opaque region (A = 1.0).
+// A fully white region (S = 1.0), on the other hand, becomes a fully transparent region (A = 0.0).
+// Values that fall within the range (0.0, 1.0) are converted into complementary alpha values (A = 1.0 - S).
+//
+// Conversion of 8bpp grayscale images is simple: Subtract the pixel's value from 255 and use the result as the alpha value.
+// Images of other bit depths have their pixels scaled to values between 0 and 255.
+static HRESULT __CGImageMaskConvertToWICAlphaBitmap(CGImageRef image, IWICBitmap** pAlphaBitmap) {
+    RETURN_HR_IF_NULL(E_INVALIDARG, image);
+    RETURN_HR_IF_NULL(E_POINTER, pAlphaBitmap);
+
+    RETURN_HR_IF(E_INVALIDARG, !CGImageIsMask(image));
+
+    ComPtr<IWICImagingFactory> imagingFactory;
+    RETURN_IF_FAILED(_CGGetWICFactory(&imagingFactory));
+
+    woc::unique_cf<CGImageRef> gray8Bitmap{ _CGImageCreateCopyWithPixelFormat(image, GUID_WICPixelFormat8bppGray) };
+    RETURN_HR_IF_NULL(E_INVALIDARG, gray8Bitmap);
+
+    ComPtr<IWICBitmap> gray8Source;
+    RETURN_IF_FAILED(_CGImageGetWICImageSource(gray8Bitmap.get(), &gray8Source));
+
+    unsigned int w = 0, h = 0;
+    RETURN_IF_FAILED(gray8Source->GetSize(&w, &h));
+
+    ComPtr<IWICBitmapLock> gray8Lock;
+    RETURN_IF_FAILED(gray8Source->Lock(nullptr, WICBitmapLockRead, &gray8Lock));
+
+    unsigned char* gray8Data;
```

Comment:
That sacrifices the locality of something else (alpha8 image stuff.)

---
In reply to: [92092724](https://github.com/Microsoft/WinObjC/pull/1528#discussion_r92092724) [](ancestors = 92092724)